src/lib/tls/asio/asio_stream.h

```diff
@@ -0,0 +1,583 @@
+       */
+      template<typename VerifyCallback>
+      void set_verify_callback(VerifyCallback callback,
+                               boost::system::error_code& ec)
+         {
+         BOTAN_UNUSED(callback);
+         ec = Botan::TLS::error::not_implemented;
+         }
+
+      /**
+       * @throws Not_Implemented
+       */
+      void set_verify_depth(int depth)
+         {
+         BOTAN_UNUSED(depth);
+         throw Not_Implemented("set_verify_depth is not implemented");
+         }
+
+      /**
+       * Not Implemented.
+       * @param ec Will be set to `Botan::TLS::error::not_implemented`
+       */
+      void set_verify_depth(int depth,
+                            boost::system::error_code& ec)
+         {
+         BOTAN_UNUSED(depth);
+         ec = Botan::TLS::error::not_implemented;
+         }
+
+      /**
+       * @throws Not_Implemented
+       */
+      template <typename verify_mode>
+      void set_verify_mode(verify_mode v)
+         {
+         BOTAN_UNUSED(v);
+         throw Not_Implemented("set_verify_mode is not implemented");
+         }
+
+      /**
+       * Not Implemented.
+       * @param ec Will be set to `Botan::TLS::error::not_implemented`
+       */
+      template <typename verify_mode>
+      void set_verify_mode(verify_mode v,
+                           boost::system::error_code& ec)
+         {
+         BOTAN_UNUSED(v);
+         ec = Botan::TLS::error::not_implemented;
+         }
+
+      //
+      // -- -- handshake methods
+      //
+
+      /**
+       * Performs SSL handshaking.
+       * The function call will block until handshaking is complete or an error occurs.
+       * @throws boost::system::system_error if error occured
+       */
+      void handshake()
+         {
+         boost::system::error_code ec;
+         handshake(ec);
+         boost::asio::detail::throw_error(ec, "handshake");
+         }
+
+      /**
+       * Performs SSL handshaking.
+       * The function call will block until handshaking is complete or an error occurs.
+       * @param ec Set to indicate what error occurred, if any.
+       */
+      void handshake(boost::system::error_code& ec)
+         {
+         while(!native_handle()->is_active())
+            {
+            sendPendingEncryptedData(ec);
+            if(ec)
+               { return; }
+
+            boost::asio::const_buffer read_buffer
+               {
+               this->m_core.input_buffer.data(),
+               m_nextLayer.read_some(this->m_core.input_buffer, ec)
+               };
+
+            if(ec)
+               { return; }
+
+            try
+               {
+               native_handle()->received_data(static_cast<const uint8_t*>(read_buffer.data()),
+                                              read_buffer.size());
+               }
+            catch(const std::exception& ex)
+               {
+               ec = Botan::TLS::convertException();
```

Comment:
Ok, we now catch both `Botan::TLS::TLS_Exception` and `Botan::Exception` wherever we used `convertException` before. The error_code this creates will either be from the `BotanAlertCategory` or the `BotanErrorCategory` and will have values from `Botan::TLS::Alert` or `Botan::ErrorType`, respectively. I hope that's the correct way of handling this.

We are still also catching `std::exception` because our code should not throw if the user expects it to set an error_code. It introduces a lot of try-catch blocks. Should we re-introduce `convertException` as a helper function that only does the throwing-and-catching again, or don't you mind the code duplication?